### DISCUSSION TITLE
equivalent of systemC wait( time ) or SystemVerilog #time

### DISCUSSION THREAD

**Discussion by AdamRose66** [intel/rohd#449] [General]
I notice that the simulator documentation for registerAction says this:

  /// Registers an abritrary [action] to be executed at [timestamp] time.
  ///
  /// The [action], if it returns a [Future], will be `await`ed.
  static void registerAction(int timestamp, dynamic Function() action) { ...

So I wondering if it is possible to write a wait( int delay ) function which awaits a Future<void> to provide wait( time ) like semantics ? Or has this already been done somewhere ?

Thanks,
Adam.

**Comment by mkorbel1** [comment]
Thanks for the question!

I think you could achieve this with something like this:
```dart
Future<void> wait(int delay) async {
  final toComplete = Completer();
  Simulator.registerAction(Simulator.time + delay, toComplete.complete);
  await toComplete.future;
}
```

Now you can just `await wait(10)` or `wait(10).then(...`, for example.

This is a useful utility, and I think something like it should be added to ROHD's `Simulator` by default.

I filed an issue tracking it here: https://github.com/intel/rohd/issues/450

**Comment by AdamRose66** [comment]
minor naming comment : if we want to expose the await syntax, then maybe `await delay(10)` or `delay(10).then(...` ?

optionally we could also have 

```
wait( int n ) async
{
  await delay( 10 );
}
```

**Reply by mkorbel1** [reply]
By the way, ROHD-VF already has a `waitCycles` function, which is useful if you wish to wait on edges rather than simulator time
https://intel.github.io/rohd-vf/rohd_vf/LogicWaiter/waitCycles.html

**Comment by AdamRose66** [comment]
Yes, I think I saw that in my search through the docs.Sometimes we don’t have an explicit clock. But my main concern is actually performance. I would hope that waiting for the time equivalent of 10,000 clock cycles should be much efficient than actually waiting on 10,000 individual clock edges.On 3 Jan 2024, at 17:36, Max Korbel ***@***.***> wrote:﻿
By the way, ROHD-VF already has a waitCycles function, which is useful if you wish to wait on edges rather than simulator time
https://intel.github.io/rohd-vf/rohd_vf/LogicWaiter/waitCycles.html

**Reply by mkorbel1** [reply]
Yes, you're correct that waiting for the equivalent time is much faster than waiting for that many clock edges.
